### PR TITLE
src: wait for threads to start

### PR DESCRIPTION
**Pull request by bnoordhuis**
Alternative to #1637, the logic in that PR didn't quite work.  cc @bzoz

CI: https://ci.nodejs.org/job/libuv-test-commit/585/

### CODE REVIEW

File: src/threadpool.c [hunk 1]
```diff
@@ -161,7 +160,11 @@ static void init_threads(void) {
     if (uv_thread_create(threads + i, worker, NULL))
       abort();
 
-  initialized = 1;
+  do {
+    uv_mutex_lock(&mutex);
+    spin = (idle_threads < nthreads);
+    uv_mutex_unlock(&mutex);
+  } while (spin);
```
Comment (bnoordhuis) [inline src/threadpool.c]: Spinning like this is admittedly not very nice but it only happens once and only for microseconds.

**Comment by santigimeno** [comment]
Why is it not working with condition variables?

**Comment by bnoordhuis** [comment]
The workers wait, they don't signal, when they increment `idle_threads`.

**Comment by santigimeno** [comment]
Sorry @bnoordhuis, I'm not sure I've expressed myself correctly. I was wondering why the solution proposed in https://github.com/libuv/libuv/pull/1637 wasn't working correctly.

**Comment by bnoordhuis** [comment]
It seems to be prone to deadlock.  The do/while should perhaps have been a while statement.

While looking at it, I figured there was probably a simpler way to do it.

File: src/threadpool.c (hunk 1 above)
Comment (bzoz) [reply src/threadpool.c]: Isn't it possible for one of the threads to pickup a work item and decrement `idle_threads`?

**Comment by santigimeno** [comment]
Ok. I think I understand now. Maybe moving this lock https://github.com/janeasystems/libuv/blob/47c90c99d24f2b302c800d91e535cb6eef255768/src/threadpool.c#L180 up some lines before creating the threads could do the trick?

File: src/threadpool.c (hunk 1 above)
Comment (bnoordhuis) [reply src/threadpool.c]: No.  init_threads() runs exclusively, it's called from the `uv_once()` in `uv__work_submit()`.

Comment (bzoz) [reply src/threadpool.c]: Right. LGTM then!

**Comment by cjihrig** [comment]
@bnoordhuis this appears to be causing problems in Node on Windows.

Compare this CI run https://ci.nodejs.org/job/node-test-binary-windows/13192/ that includes this commit with this one https://ci.nodejs.org/job/node-test-commit-windows-fanned/13854/ where I reverted this commit.

**Comment by bnoordhuis** [comment]
@cjihrig I did a run with just that commit on top of master and it's green: https://ci.nodejs.org/job/node-compile-windows/13915/

Maybe there is some weak interaction going on with another commit?  Or maybe the CI was just in bad mood.

**Comment by bzoz** [comment]
Just tried with and without this commit, everything works great for me locally. I guess it was a CI fluke.

**Comment by cjihrig** [comment]
Thank you both for looking into it. I just kicked off another CI run for the 1.17.0 update in Node.

https://ci.nodejs.org/job/node-test-pull-request/11799/

**Comment by cjihrig** [comment]
Hm, it looks like the new CI run is still failing. Also, @bnoordhuis you linked to the Windows compile job. Is this the corresponding test run - https://ci.nodejs.org/job/node-test-binary-windows/13232/? If so, it looks like it had the same issues I saw.

**Comment by bnoordhuis** [comment]
Ah, can't believe I got that wrong.  Well, the proof is in the pudding, and the revert is in https://github.com/libuv/libuv/pull/1656.

**Comment by davisjam** [comment]
I think *some* working version of this PR should go in. Seems like a nice addition for reasoning about the threadpool.

I'm nervous about using idle_threads as the counter, because libuv docs warn that "Callers should be prepared to deal with spurious wakeups on uv_cond_wait() and uv_cond_timedwait()." The value of idle_threads could in principle be untrustworthy, and might never reach nthreads.

Can you try with a semaphore? Each worker can `post` before entering the loop, and `init_threads()` can `wait` once for each worker.

**Comment by bnoordhuis** [comment]
> I'm nervous about using idle_threads as the counter, because libuv docs warn that "Callers should be prepared to deal with spurious wakeups on uv_cond_wait() and uv_cond_timedwait()." The value of idle_threads could in principle be untrustworthy, and might never reach nthreads.

I don't follow.  `worker()` goes to sleep again on a spurious wake-up because of this loop:
```c
    uv_mutex_lock(&mutex);

    while (QUEUE_EMPTY(&wq)) {
      idle_threads += 1;
      uv_cond_wait(&cond, &mutex);
      idle_threads -= 1;
    }
```
> Can you try with a semaphore?

I tried to avoid using one because they're fairly heavy-weight objects on some platforms.  It seemed a bit wasteful for something that is only used once.

**Comment by bzoz** [comment]
I tried my https://github.com/libuv/libuv/pull/1637 with mutex moved before the loop (https://github.com/janeasystems/node/commit/d224eab47764a9c363d53eadff06e683352647d6), it passes CI: https://ci.nodejs.org/job/node-test-commit-windows-fanned/13917/

**Comment by davisjam** [comment]
> I don't follow. worker() goes to sleep again on a spurious wake-up because of this loop

Right, my mistake. The issue is actually that an infinite supply of spurious wakeups could prevent anyone else from acquiring the mutex, probably needlessly paranoid unless the threadpool is enormous.

> I tried to avoid using one because they're fairly heavy-weight objects on some platforms. It seemed a bit wasteful for something that is only used once.

Since we're already paying the cost of starting a bunch of threads and then waiting for them to make some progress, do you think a semaphore will cost noticeably extra? Premature optimization?

**Comment by bnoordhuis** [comment]
I was explaining why I didn't use one, not saying we shouldn't use one now.

**Comment by brandonpoc** [comment]
Sorry to drag this thing back alive - but quick question; what does it mean by "spurious wakeups"? Am I to not rely on a conditional wait only proceeding on a condition signalling from another thread, or does this mean something else?

**Pull request closed without merging**